### ISSUE TITLE
howto: generate tables with captions

### ISSUE THREAD

**Issue by rriemann** [htdebeer/paru#40]
Hello,

I wrote a filter that replaces some `::paru` paragraph with a markdown table. Preparing the table as markdown does not feel like a very clean solution.

Is there a way to let paru directly convert a ruby 2D array/Hash to a table? Is there support to set a caption?

Best,
Robert

**Comment by htdebeer** [comment]
There is no such functionality at the moment. But some convenience methods or class could be nice.

One way to go about is to create a PandocTable wrapper class you can initialize or fill with an array or hash, and set the caption. This wrapper class would have a method "to_markdown" that generates pandoc compatible markdown you can use to replace/add to the AST. It would also be nice to get such a PandocTable object from a Table node in the AST for easy manipulation of the table.

Another way to go about it, could be to add utility/convenience methods to set/get the Table node as/with an Array or Hash. 

Either way, do note that the contents of the cells and caption are AST nodes themselves as well. They could contain complex data.

Can you create an example scenario of what exactly you're trying to accomplish? Just to get a feel of how you would like this addition to the API to look and feel.

For example: How would the 2D array look like? How would the Hash work? What's the contents of the nodes in the Hash or Array? Should this functionality be two-way: from AST to Hash/Array and Array/Hash to AST?

**Comment by rriemann** [comment]
So far, I have data in Hash format (I just thought 2D array would be easier to deal with).

```ruby
yaml = [{:x=>0, :y=>0}, {:x=>1, :y=>2}, {:x=>2, :y=>4}, {:x=>3, :y=>6}, {:x=>4, :y=>8}, {:x=>5, :y=>10}, {:x=>6, :y=>12}, {:x=>7, :y=>14}, {:x=>8, :y=>16}, {:x=>9, :y=>18}]
```

I have a function to build a markdown table from this kind of data:


```ruby
def build_table(yaml, columns)

  keys = if columns
    columns.split(',').map(&:to_sym)
  else
    yaml.first.keys
  end
  table  = '| ' + keys.map(&:to_s).join(' | ') + " |\n"
  table += '| ' + keys.size.times.map{' :-- '}.join(' | ') + " |\n"
  yaml.each do |set|
    row = keys.map do |key|
      cell = set[key].to_s
      # autolink
      if cell =~ URI::regexp
        "<#{cell}>"
      else
        cell
      end
    end
    table += '| ' + row.join(' | ') + " |\n"
  end
  table += "\n"

  return table
end
```

Then, I can call in a paru filter:

`paru_element.outer_markdown = build_table(yaml)`

However, I believe it would be better to directly build an AST from the hash/2D table.

Problems with this approach:

- I cannot control the column widths
- caption is not yet integrated

**Comment by htdebeer** [comment]
Thanks for your input! 

Some sort of "rich table component" would be nice. Adding a caption shouldn't be too hard. But how would imagine controlling column's widths? 

And how to treat markdown content in cells? We could call pandoc to convert each cell's content, but that's a bit expensive performance wise.

Once we start with this, we can extend it even further, for example with CSV input/output.

**Comment by htdebeer** [comment]
Now that I am thinking about it, convenience methods/objects could be useful for other more complex elements in the Pandoc AST, such as:

- Lists that can import/export Arrays
- Definition lists that can import/export Hashes
- Code blocks that can import/export Strings or Files

Maybe more?

**Comment by rriemann** [comment]
Column widths should be either not set (reverting to automatic for some sorts of exports such as docx if I am not mistaken) or set manually.

In my current project, I focus on tex output and need to set this manually in any case.

My data is not in markdown format, but plain text. However, some convenience ruby filters (methods) would be nice to convert plain text link to an actual link, use lists and emphasis.

One of these ruby filters could even call pandoc for those that want more.

**Comment by htdebeer** [comment]
All right. I'll look into this a bit further in the coming weeks. I do like the idea of enhancing the paru API, to make it more convenient than the bare-bones filter API offered by pandoc.

**Comment by htdebeer** [comment]
To anyone reading this conversation, if you have more wishes or ideas to make the paru API more convenient that are not related to tables, please submit separate requests and tag it with the "convenience API" tag.

**Comment by htdebeer** [comment]
I've just published an alpha version of paru 0.3.1.0.a. It contains convenience methods for:

* CodeBlock: to_file, from_file, to_code_string, from_code_string
* OrderedList: to_array, from_array
* BulletList: to_array, from_array
* DefinitionList: to_array, from_array
* Table: to_file, from_file, to_array, from_array

See the [filter_test](https://github.com/htdebeer/paru/blob/master/test/test_filter.rb) for an example how to use these. There is also some [api documentation](https://heerdebeer.org/Software/markdown/paru/documentation/api-doc/Paru/PandocFilter/Table.html).

Feedback, suggestions, and improvements are welcome. Real-life examples as well.

**Comment by htdebeer** [comment]
No reply; closing this issue.
